### DISCUSSION TITLE
TypeScript issue when using `defer`, `return redirect()` and `SerializeFrom<typeof loader>`

### DISCUSSION THREAD

**Discussion by kevinreber** [remix-run/remix#7720] [Q&A] [answered]
## Summary
We have a `loader` with a `defer` that has some logic to `redirect()` a user if they are not logged in. At the same time we're using `SerializeFrom<typeof loader` to help abstract the types from our `loader` (snippet below) 

_Note: This seems to work fine when using `json` over `defer`_

```tsx
export type LoaderTypes = SerializeFrom<typeof loader>;

export async function loader({ request }: LoaderArgs) {
  const isUserAuthenticated = await authenticateUser();
  if (!isUserAuthenticated) {
    return redirect("/login");
  }

  return defer({ data: "Some data" });
}

export default function Index() {
  const loaderData = useLoaderData<LoaderTypes>();
  <React.Suspense fallback={<p>Loading...</p>}>
    <Await resolve={loaderData.data}>
      {(data) => {
        return <h1>{data}</h1>;
      }}
    </Await>
  </React.Suspense>;
}
```

When excluding `return redirect()`, our `LoaderTypes` look like this
```tsx
type LoaderTypes = { data: string }
```
When adding the logic for `return redirect()`, our `LoaderTypes` look like this
```tsx
type LoaderTypes = SerializeObject<UndefinedToOptional<TypedResponse<never>>> | SerializeDeferred<{
    data: string;
}>
```

which leads us to the following TypeScript error
```
Property 'data' does not exist on type 'SerializeObject<UndefinedToOptional<SerializeObject<UndefinedToOptional<TypedResponse<never>>>>> | SerializeObject<...>'.
  
 Property 'data' does not exist on type 'SerializeObject<UndefinedToOptional<SerializeObject<UndefinedToOptional<TypedResponse<never>>>>>'.
```
<img width="1172" alt="CleanShot 2023-10-20 at 13 58 53@2x" src="https://github.com/remix-run/remix/assets/42260999/25068c86-3c34-4d3f-bfb5-44097800770c">

Has anyone else ever come across this issue?

## References
- Stackblitz Repo: https://stackblitz.com/edit/remix-run-remix-uutlwq?file=app%2Froutes%2F_index.tsx

**Comment by kiliman** [comment, marked as answer]
If you want TypeScript to infer types, don't `return redirect`. Instead, you should `throw redirect`.

The reason is that `redirect` returns a generic `Response` that can't be serialized to JSON. Whereas the `json` and `defer` helpers do return an object that will be serialized. Remix has special types that handle this conversion/inference for you.

This is where the magic happens: https://github.com/remix-run/remix/blob/main/packages/remix-server-runtime/responses.ts
